File: clients/android/app/src/main/kotlin/com/lextures/android/core/lms/OrgStructureAdminLogic.kt
package com.lextures.android.core.lms

import com.lextures.android.core.navigation.MobilePlatformFeatures
import com.lextures.android.core.network.ApiError
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.TimeZone

/** Organizations, org units, and academic terms admin helpers (M14.4). */
object OrgStructureAdminLogic {
    const val RBAC_MANAGE_PERMISSION = "global:app:rbac:manage"
    const val ORG_UNITS_ADMIN_PERMISSION = "tenant:org:units:admin"
    const val DEFAULT_TERM_TYPE = "semester"
    const val ORG_LIST_LIMIT = 200

    private val isoDateFormat: SimpleDateFormat by lazy {
        SimpleDateFormat("yyyy-MM-dd", Locale.US).apply {
            timeZone = TimeZone.getTimeZone("UTC")
        }
    }

    fun adminSettingsEnabled(features: MobilePlatformFeatures): Boolean =
        features.ffMobileAdminSettings

    fun canManageOrganizations(permissions: List<String>): Boolean =
        permissions.contains(RBAC_MANAGE_PERMISSION)

    fun canManageOrgUnitsAndTerms(permissions: List<String>): Boolean =
        permissions.contains(RBAC_MANAGE_PERMISSION) ||
            permissions.contains(ORG_UNITS_ADMIN_PERMISSION)

    fun shouldShowEntry(
        features: MobilePlatformFeatures,
        permissions: List<String>,
    ): Boolean =
        adminSettingsEnabled(features) &&
            (canManageOrganizations(permissions) || canManageOrgUnitsAndTerms(permissions))

    fun canView(
        features: MobilePlatformFeatures,
        permissions: List<String>,
    ): Boolean = shouldShowEntry(features, permissions)

    fun webOrganizationsPath(): String = "/settings/organizations"
    fun webOrgUnitsPath(): String = "/settings/org-units"
    fun webTermsPath(): String = "/settings/terms"

    fun resolveOrgId(accessToken: String?, courses: List<CourseSummary>): String? =
        CourseCreateLogic.resolveOrgId(accessToken, courses)

    fun normalizedName(value: String): String = value.trim()

    fun isValidTermName(value: String): Boolean = normalizedName(value).isNotEmpty()

    fun isoDateString(date: Date): String = isoDateFormat.format(date)

    fun dateFromIso(value: String?): Date? {
        val trimmed = value?.trim().orEmpty()
        if (trimmed.isEmpty()) return null
        return runCatching { isoDateFormat.parse(trimmed) }.getOrNull()
    }

    fun formatDateRange(start: String?, end: String?): String {
        val startText = start?.takeIf { it.isNotBlank() } ?: "—"
        val endText = end?.takeIf { it.isNotBlank() } ?: "—"
        return "$startText — $endText"
    }

    fun isValidDateRange(start: String, end: String): Boolean {
        val startDate = dateFromIso(start) ?: return false
        val endDate = dateFromIso(end) ?: return false
        return !endDate.before(startDate)
    }

    fun createTermRequest(
        name: String,
        termType: String,
        startDate: String,
        endDate: String,
    ): CreateAcademicTermRequest =
        CreateAcademicTermRequest(
            name = normalizedName(name),
            termType = termType.ifBlank { DEFAULT_TERM_TYPE },
            startDate = startDate,
            endDate = endDate,
        )

    fun patchTermDatesRequest(startDate: String, endDate: String): PatchAcademicTermRequest =
        PatchAcademicTermRequest(startDate = startDate, endDate = endDate)

    fun patchOrgUnitNameRequest(name: String): PatchOrgUnitRequest =
        PatchOrgUnitRequest(name = normalizedName(name))

    fun flattenTree(nodes: List<OrgUnitTreeNode>): List<OrgUnitTreeNode> =
        nodes.flatMap { node -> listOf(node) + flattenTree(node.children.orEmpty()) }

    fun userFacingError(error: Throwable, genericMessage: String): String =
        (error as? ApiError.HttpStatus)?.message?.takeIf { it.isNotEmpty() } ?: genericMessage
}

File: clients/ios/Lextures/Core/LMS/OrgStructureAdminLogic.swift
import Foundation

/// Organizations, org units, and academic terms admin helpers (M14.4).
enum OrgStructureAdminLogic {
    static let rbacManagePermission = "global:app:rbac:manage"
    static let orgUnitsAdminPermission = "tenant:org:units:admin"
    static let defaultTermType = "semester"
    static let orgListLimit = 200

    static func adminSettingsEnabled(_ features: MobilePlatformFeatures) -> Bool {
        features.ffMobileAdminSettings || features.ffMobileAdminConsole
    }

    static func canManageOrganizations(permissions: [String]) -> Bool {
        permissions.contains(rbacManagePermission)
    }

    static func canManageOrgUnitsAndTerms(permissions: [String]) -> Bool {
        permissions.contains(rbacManagePermission) || permissions.contains(orgUnitsAdminPermission)
    }

    static func shouldShowEntry(
        features: MobilePlatformFeatures,
        permissions: [String]
    ) -> Bool {
        !features.ffMobileAdminConsole
            && features.ffMobileAdminSettings
            && (canManageOrganizations(permissions: permissions)
                || canManageOrgUnitsAndTerms(permissions: permissions))
    }

    static func canView(
        features: MobilePlatformFeatures,
        permissions: [String]
    ) -> Bool {
        adminSettingsEnabled(features)
            && (canManageOrganizations(permissions: permissions)
                || canManageOrgUnitsAndTerms(permissions: permissions))
    }

    static func webOrganizationsPath() -> String { "/settings/organizations" }
    static func webOrgUnitsPath() -> String { "/settings/org-units" }
    static func webTermsPath() -> String { "/settings/terms" }

    static func resolveOrgId(accessToken: String?, courses: [CourseSummary]) -> String? {
        CourseCreateLogic.resolveOrgId(accessToken: accessToken, courses: courses)
    }

    static func normalizedName(_ value: String) -> String {
        value.trimmingCharacters(in: .whitespacesAndNewlines)
    }

    static func isValidTermName(_ value: String) -> Bool {
        !normalizedName(value).isEmpty
    }

    static func isoDateString(from date: Date) -> String {
        let formatter = DateFormatter()
        formatter.calendar = Calendar(identifier: .gregorian)
        formatter.locale = Locale(identifier: "en_US_POSIX")
        formatter.timeZone = TimeZone(secondsFromGMT: 0)
        formatter.dateFormat = "yyyy-MM-dd"
        return formatter.string(from: date)
    }

    static func date(fromIso value: String?) -> Date? {
        guard let value, !value.isEmpty else { return nil }
        let formatter = DateFormatter()
        formatter.calendar = Calendar(identifier: .gregorian)
        formatter.locale = Locale(identifier: "en_US_POSIX")
        formatter.timeZone = TimeZone(secondsFromGMT: 0)
        formatter.dateFormat = "yyyy-MM-dd"
        return formatter.date(from: value)
    }

    static func formatDateRange(start: String?, end: String?) -> String {
        let startText = start ?? "—"
        let endText = end ?? "—"
        return "\(startText) — \(endText)"
    }

    static func isValidDateRange(start: String, end: String) -> Bool {
        guard let startDate = date(fromIso: start), let endDate = date(fromIso: end) else {
            return false
        }
        return endDate >= startDate
    }

    static func createTermRequest(
        name: String,
        termType: String,
        startDate: String,
        endDate: String
    ) -> CreateAcademicTermRequest {
        CreateAcademicTermRequest(
            name: normalizedName(name),
            termType: termType.isEmpty ? defaultTermType : termType,
            startDate: startDate,
            endDate: endDate
        )
    }

    static func patchTermDatesRequest(startDate: String, endDate: String) -> PatchAcademicTermRequest {
        PatchAcademicTermRequest(
            name: nil,
            termType: nil,
            startDate: startDate,
            endDate: endDate,
            status: nil
        )
    }

    static func patchOrgUnitNameRequest(name: String) -> PatchOrgUnitRequest {
        PatchOrgUnitRequest(name: normalizedName(name))
    }

    static func flattenTree(_ nodes: [OrgUnitTreeNode]) -> [OrgUnitTreeNode] {
        nodes.flatMap { node in
            [node] + flattenTree(node.children ?? [])
        }
    }

    static func userFacingError(_ error: Error) -> String {
        if let apiError = error as? APIError, case let .httpStatus(_, message) = apiError,
           let message, !message.isEmpty {
            return message
        }
        return L.text("mobile.admin.orgStructure.error")
    }
}
